include/libvmm/virtio/net.h

```diff
@@ -0,0 +1,275 @@
+ * byte VLAN ID.  VLAN filterting is available with the
+ * VIRTIO_NET_F_CTRL_VLAN feature bit.
+ */
+#define VIRTIO_NET_CTRL_VLAN       2
+#define VIRTIO_NET_CTRL_VLAN_ADD             0
+#define VIRTIO_NET_CTRL_VLAN_DEL             1
+
+/*
+ * Control link announce acknowledgement
+ *
+ * The command VIRTIO_NET_CTRL_ANNOUNCE_ACK is used to indicate that
+ * driver has recevied the notification; device would clear the
+ * VIRTIO_NET_S_ANNOUNCE bit in the status field after it receives
+ * this command.
+ */
+#define VIRTIO_NET_CTRL_ANNOUNCE       3
+#define VIRTIO_NET_CTRL_ANNOUNCE_ACK         0
+
+/*
+ * Control Receive Flow Steering
+ *
+ * The command VIRTIO_NET_CTRL_MQ_VQ_PAIRS_SET
+ * enables Receive Flow Steering, specifying the number of the transmit and
+ * receive queues that will be used. After the command is consumed and acked by
+ * the device, the device will not steer new packets on receive virtqueues
+ * other than specified nor read from transmit virtqueues other than specified.
+ * Accordingly, driver should not transmit new packets  on virtqueues other than
+ * specified.
+ */
+struct virtio_net_ctrl_mq {
+    uint16_t virtqueue_pairs;
+};
+
+#define VIRTIO_NET_CTRL_MQ   4
+#define VIRTIO_NET_CTRL_MQ_VQ_PAIRS_SET        0
+#define VIRTIO_NET_CTRL_MQ_VQ_PAIRS_MIN        1
+#define VIRTIO_NET_CTRL_MQ_VQ_PAIRS_MAX        0x8000
+
+/* Backend implementation */
+#define SDDF_NET_NUM_HANDLES    3
+#define SDDF_NET_RX_HANDLE      0
+#define SDDF_NET_TX_HANDLE      1
+/* @jade: we need a channel for getmac protocol, however, channel numbers are
+ * associated with the handler structure (is it what we want?). I am not sure
+ * if we should have a dummy handler for getmac since getmac doesn't require
+ * any SDDF queue. */
+#define SDDF_NET_GETMAC_HANDLE  2
+
+#define VIRTIO_NET_NUM_VIRTQ    2
+#define VIRTIO_NET_RX_VIRTQ     0
+#define VIRTIO_NET_TX_VIRTQ     1
+
+// @jade, @ericc: These are sDDF specific, belong in a configuration file elsewhere ideally
+#define SHMEM_NUM_BUFFERS 512
+#define SHMEM_BUF_SIZE 2048
+
+// @jade: I don't know why we need these but the driver seems to care
+typedef enum {
+    ORIGIN_VIRTIO_NET_RX_VIRTQ,
+    ORIGIN_TX_QUEUE,
+} ethernet_buffer_origin_t;
+
+typedef struct ethernet_buffer {
+    /* The acutal underlying memory of the buffer */
+    uintptr_t buffer;
+    /* The physical size of the buffer */
+    size_t size;
+    /* Queue from which the buffer was allocated */
+    char origin;
+    /* Index into buffer_metadata array */
+    unsigned int index;
+    /* in use */
+    bool in_use;
+} ethernet_buffer_t;
+
+struct virtio_net_device {
+    struct virtio_device virtio_device;
+
+    struct virtio_net_config config;
+    struct virtio_queue_handler vqs[VIRTIO_NET_NUM_VIRTQ];
+
+    net_queue_handle_t rx;
+    net_queue_handle_t tx;
+    void *rx_data;
+    void *tx_data;
+    int tx_ch;
+    int rx_ch;
+
+    ethernet_buffer_t buffer_metadata[SHMEM_NUM_BUFFERS * 2];
```

Comment:
Hmm, I can't find any uses of `buffer_metadata` or `ethernet_buffer_t`. Is this used anywhere?

Reply:
I've sent a message to @abrandnewusername about this. I assume it is to keep track of active and free buffers similar to sound and block but I can't find anything in the virtIO spec